src/nvim/api/buffer.c

```diff
@@ -738,6 +742,223 @@ ArrayOf(Integer, 2) nvim_buf_get_mark(Buffer buffer, String name, Error *err)
+                                  l_lnum,
+                                  l_col,
+                                  u_lnum,
+                                  u_col,
+                                  (int64_t)amount,
+                                  reverse ? BACKWARD: FORWARD);
+
+  size_t n = kv_size(extmarks_in_range);
+  for (size_t i = 0; i < n; i++) {
+    mark.size = 0;
+    mark.capacity = 0;
+    mark.items = 0;
+    extmark = kv_A(extmarks_in_range, i);
+    ADD(mark, INTEGER_OBJ((Integer)extmark->mark_id));
+    ADD(mark, INTEGER_OBJ(extmark->line->lnum));
+    ADD(mark, INTEGER_OBJ(extmark->col));
+    ADD(rv, ARRAY_OBJ(mark));
+  }
+  kv_destroy(extmarks_in_range);
+  return rv;
+
+}
+
+/// Create or update an extended mark at a position
+///
+/// If an invalid namespace or buffer is given, an error will be
+/// raised and 0 returned.
+///
+/// @param buffer The buffer handle
+/// @param namespace a identifier returned previously with nvim_create_namespace
+/// @param id The extended mark's id or 0 for a randomly generated id
+/// @param row The row to set the extended mark to.
+/// @param col The column to set the extended mark to.
+/// @param[out] err Details of an error that may have occurred
+/// @return 1 on new, 2 on update; or a mark_id if argument mark_id was 0
+Integer nvim_buf_set_mark(Buffer buffer,
+                          Integer namespace,
+                          Integer mark_id,
+                          Integer row,
+                          Integer col,
+                          Error *err)
+    FUNC_API_SINCE(3)
+{
+  Integer rv = 0;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  if (!buf) {
+    return rv;
+  }
+  if (!ns_initialized((uint64_t)namespace)) {
+    api_set_error(err, kErrorTypeValidation, _("Invalid mark namespace"));
+    return rv;
+  }
+  if (row < 1 || col < 1) {
+    api_set_error(err, kErrorTypeValidation, _("Row and column must be greater than 0"));
+    return rv;
+  }
+
+  bool return_id = false;
+  uint64_t id;
+  if (mark_id == 0) {
+    id = extmark_free_id_get(buf, (uint64_t)namespace);
+    return_id = true;
+  } else if (mark_id > 0) {
+    id = (uint64_t)mark_id;
+  } else {
+    api_set_error(err, kErrorTypeValidation, _("Invalid mark id"));
+    return rv;
+  }
+
+  rv = (Integer)extmark_set(buf, (uint64_t)namespace, id,
+                            (linenr_T)row, (colnr_T)col, kExtmarkUndo);
+  if (return_id) {
+    return (Integer)id;
+  } else {
+    return rv;
+  }
+}
+
+/// Remove an extended mark
+///
+/// If an invalid namespace or buffer is given, an error will be
+/// raised and 0 returned.
+///
+/// @param buffer The buffer handle
+/// @param namespace a identifier returned previously with nvim_create_namespace
+/// @param id The extended mark's id
+/// @param[out] err Details of an error that may have occurred
+/// @return 1 on success, 0 on no mark found
+Integer nvim_buf_unset_mark(Buffer buffer,
```

Comment:
See `:help dev-api` for function naming guidelines.

Currently in the API we have `clear` as a verb (`nvim_buf_clear_highlight`) and `del`. I don't think we need `unset` or `remove`, so pick one of `del` or `clear`.

    nvim_buf_del_mark

Reply:
`mark` or `extmark`?